lib/infrastructure/plan/plan_data_source.dart

```diff
@@ -40,4 +46,59 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
     }
     return;
   }
+
+  Future<List<String>> getBookmarkedPlanIds() async {
+    final userData = fireauth.currentUser;
+    if (userData == null) {
+      return [];
+    }
+    final snapshot =
+        await firestore.collection('users').doc(userData.uid).get();
+    if (!snapshot.exists ||
+        !snapshot.data()!.containsKey('bookmarkedPlanIds')) {
+      return [];
+    }
+    final bookmarkedPlanIds = snapshot.data()!['bookmarkedPlanIds'];
+    if (bookmarkedPlanIds == null || bookmarkedPlanIds is! List<dynamic>) {
+      return [];
+    }
+    return List<String>.from(bookmarkedPlanIds);
+  }
+
+  Future<Plan> getPlanData({required String planId}) async {
+    final topics = <Topic>[];
+    final snapshot =
+        await firestore.collection('PopularPlans').doc(planId).get();
+    final data = snapshot.data();
+    if (data == null) {
+      return const Plan(
+        title: '',
+        description: '',
+        thumbnailUrl: '',
+        topics: [],
+      );
+    }
+    for (final topicId in data['topics'] as List) {
+      final topic = await topicsorce.getTopicData(topicId: topicId.toString());
+      topics.add(Topic(name: topic.name, thumbnailUrl: ''));
+    }
+    return Plan(
+      id: planId,
+      title: data['title'].toString(),
+      description: data['description'].toString(),
+      thumbnailUrl: data['thumbnailUrl'].toString(),
+      topics: topics,
+    );
+  }
```

Comment:
ここも他のメソッドの書き方を真似しましょう
直接Map<String, dynamic>の型から文字列でデータを抽出している部分はないはず
fromJsonメソッドというfirestoreから受け取ったデータをDartの型に直す便利なメソッドがあるので、使いましょう

```suggestion
    final res = Plan.fromJson(doc.data());
    return res.copyWith(
      topics: topics,
    );
  }
```